### PR TITLE
issue #3805 - fix FHIRPath evaluator string function error handling

### PR DESCRIPTION
**Pull request by PrasannaHegde1**
Signed-off-by: Prasanna Hegde <prasanna.hegde1@ibm.com>

### CODE REVIEW

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ToStringFunction.java [hunk 1]
```diff
@@ -38,9 +38,7 @@ public int getMaxArity() {
 
     @Override
     public Collection<FHIRPathNode> apply(EvaluationContext evaluationContext, Collection<FHIRPathNode> context, List<Collection<FHIRPathNode>> arguments) {
-        if (!isSingleton(context)) {
-            return empty();
-        }
+        checkStringValue(context);
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/ToStringFunction.java:41]: toString doesn't exist under http://hl7.org/fhirpath/N1/#string-manipulation and so shouldn't use this check.  

instead this one falls under section 5.5 which says this:
> The functions in this section operate on collections with a single item. If there is more than one item, the evaluation of the expression will end and signal an error to the calling environment.

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ContainsFunction.java [hunk 2]
```diff
@@ -36,9 +35,7 @@ public int getMaxArity() {
     
     @Override
     public Collection<FHIRPathNode> apply(EvaluationContext evaluationContext, Collection<FHIRPathNode> context, List<Collection<FHIRPathNode>> arguments) {
-        if (!hasStringValue(context)) {
-            return empty();
-        }
+        checkStringValue(context);
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/ContainsFunction.java:38]: Most (all?) of the function definitions say that if the input collection is empty, then the result must be empty as well (not an error).  Please make sure we're handling this case.

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ToStringFunction.java [hunk 3]
```diff
@@ -38,7 +39,10 @@ public int getMaxArity() {
 
     @Override
     public Collection<FHIRPathNode> apply(EvaluationContext evaluationContext, Collection<FHIRPathNode> context, List<Collection<FHIRPathNode>> arguments) {
-        checkSingleton(context);
+        if(!isSingleton(context)) {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/ToStringFunction.java:41]: I think this one should return empty if the input collection is empty, but I don't think it does with this

File: fhir-path/src/main/java/com/ibm/fhir/path/function/FHIRPathStringAbstractFunction.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * Abstract class which contains common apply logic for all abstract string manipulation functions
+ */
+public abstract class FHIRPathStringAbstractFunction extends FHIRPathAbstractFunction {
+    
+    /**
+     * @implSpec Common apply logic for all abstract string manipulation functions. Delegates to doApply()
+     */
+    @Override
+    public Collection<FHIRPathNode> apply(EvaluationContext evaluationContext, Collection<FHIRPathNode> context, List<Collection<FHIRPathNode>> arguments) {
+        if(context.isEmpty()) {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/FHIRPathStringAbstractFunction.java:31]: minor style thing:  in this project we always put a space between `if` and `(` to distinguish from function calls (and same for other control flow keywords `for`, `while`, etc.)

File: fhir-path/src/main/java/com/ibm/fhir/path/function/FHIRPathStringAbstractFunction.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+import java.util.Collection;
+import java.util.List;
+
+import com.ibm.fhir.path.FHIRPathNode;
+import com.ibm.fhir.path.FHIRPathType;
+import com.ibm.fhir.path.evaluator.FHIRPathEvaluator.EvaluationContext;
+
+/**
+ * Abstract class which contains common apply logic for all abstract string manipulation functions
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/FHIRPathStringAbstractFunction.java:22]: nice

File: fhir-path/src/main/java/com/ibm/fhir/path/function/ContainsFunction.java (hunk 2 above)
Comment (lmsurpre) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/ContainsFunction.java]: looks good in the new common parent; I left a separate comment for the toString function

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java [hunk 6]
```diff
@@ -1238,4 +1238,21 @@ public static com.ibm.fhir.model.type.String getDisplay(FHIRPathTree tree, FHIRP
...
         return null;
     }
+    
+    /**
+     * returns true if the node is a String or if FHIR element can be automatically converted to System.String 
+     * https://www.hl7.org/fhir/fhirpath.html#types
+     * @param node
+     * @return boolean
+     */
+    public static boolean isStringSubType(FHIRPathNode node) {
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:1248]: One thing I didn't think of on the call... we should be clear about whether this should return true or false for FHIRPathNode of type SYSTEM_STRING

Options:
A. add this type to the list of types we return true for; or
B. change the method signature to `isFHIRStringSubType(FHIRPathElementNode)` to make it clear this only operates over FHIRPathNode that wrap FHIR elements (not primitive values)

File: fhir-path/src/main/java/com/ibm/fhir/path/function/FHIRPathStringAbstractFunction.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        if (!isSingleton(context)) {
+            throw new IllegalArgumentException("Input collection must not contain more than one item, but found " + context.size());
+            
+        } else if (!hasStringValue(context)) {
+            FHIRPathNode node = getSingleton(context);
+            if (isStringSubType(node)) {
+                logger.fine("collection item does not have a value, returning empty");
+                return empty();
+            }
+            throw new IllegalArgumentException("Input collection item must be of type String, but found '"+node.type().getName()+ "'");
```
Comment (lmsurpre) [inline fhir-path/src/main/java/com/ibm/fhir/path/function/FHIRPathStringAbstractFunction.java:48]: very minor style thing
```suggestion
            throw new IllegalArgumentException("Input collection item must be of type String, but found '" + node.type().getName() + "'");
```

Comment (PrasannaHegde1) [reply fhir-path/src/main/java/com/ibm/fhir/path/function/FHIRPathStringAbstractFunction.java]: fixed the above review comment

File: fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java (hunk 6 above)
Comment (PrasannaHegde1) [reply fhir-path/src/main/java/com/ibm/fhir/path/util/FHIRPathUtil.java:1248]: Handled the above review comment with Option A.

**Review by lmsurpre — APPROVED** [review]
LGTM

**Pull request merged**